### ISSUE TITLE
Ruby: Update and improve debugging lesson

### ISSUE THREAD

**Issue by ChargrilledChook** [TheOdinProject/curriculum#24994]
### Describe your suggestion

The Ruby debugging lesson is often a source of confusion for learners. There are a number of issues with it - many concepts are glossed over, there is a lot of configuration and probably not enough guided practice. 

On top of that, the mix between using `pry-byebug` and the VSCode `rdbg` integration needs work. While learners who get the integrated debugger working tend to find a lot of value in it, the current instructions are confusing and don't provide enough context or conceptual support. 

### Possible Improvements

- There is an opportunity to streamline the content here by switching the recommended debugger from `pry` to `rdbg`. It's now mature enough, has all the same features, is more familiar (having more similar commands to irb) and has the advantage of being the offiicial debugger for both Ruby and Rails. Unless there's a really compelling reason, it makes sense to teach the standard library over a gem. It also makes sense to use both the command line version and the vscode integration of the same tool, rather than two separate ones.

- If we go with the above, we should cover using `rdbg` via both command line and VSCode. 

- There should be some more detail on basic commands  and concepts of debugging tools - such as the difference between step in, step out and step over.

- The details on configuring `rdbg`'s VSCode integration should be improved + common issues should be addressed.

### Suggested Plan

The lesson is already long. It is likely worth splitting this up into two sections. 

The first section is largely fine but needs to be a bit more streamlined / beginner friendly. It could remain as is, but have references to `pry` swapped with `binding.irb` instead. It's a much easier way to get people started with more interactive debugging, as it uses a tool they're already familiar with and doesn't require installing or `require`ing a gem.

`rdbg` / interactive debugging tools could then be covered in a new lesson.

 

### Path

Ruby / Rails

### Lesson Url

https://www.theodinproject.com/lessons/ruby-debugging

### Checks

- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [ ] Would you like to work on this issue?

### (Optional) Additional Comments

If no one is interested in taking this on, I'll do it - but I'm happy to collaborate and / or delegate parts of it.

**Comment by ChargrilledChook** [comment]
cc @TheOdinProject/rails-path

**Comment by CouchofTomato** [comment]
Hey @ChargrilledChook 

Is this something you're waiting for feedback on? Sorry for the delay mate. It's been a hectic start to the year.

**Comment by ChargrilledChook** [comment]
@JoshDevHub I wrote a decent amount of (draft) content but had to put it down. If you're interested I would love to have your help on this one; I can share what I've done so far and you can use it (or not) depending on your thoughts. 

I agree this is probably the lesson most in need of fixing at the moment because I continually see people having technical issues

**Comment by JoshDevHub** [comment]
@ChargrilledChook Yeah I'd be interested in working on it. Also definitely interested in seeing your draft, I feel like you have a lot of good knowledge with debugging.

**Comment by ChargrilledChook** [comment]
@JoshDevHub sounds great, I'll dig up my drafts (they're in a random branch somewhere). 

Do you have thoughts on my suggestions / discussion points in the body of the issue? I have some strong opinions but there's a lot of different ways we could approach it

**Comment by JoshDevHub** [comment]
@ChargrilledChook I agree with basically every point in the issue body I think. Some of my thoughts:

- Moving away from `pry` is a good idea with `rdbg` in standard lib. I think one of the justifications for still teaching `pry` was that people might get in a job on an older version of ruby, but the debug gem has compatibility all the way back to 2.6. And so long as we're still showing how to use it in the terminal (which I think is still a good idea as well), it's not like it's some ultra challenge to move between `rdbg` via CLI and `pry` anyway. It'll definitely streamline the lesson(s) a ton to just show one tool.
- *Strongly* agree that step-in, step-out, and step-over should be covered. I'm a bit surprised they're not currently mentioned. Some other stuff that'd be useful I think:
  - Conditional breakpoints: I didn't know about this for a while but I think it's very helpful to be able to say things like `binding.b if @ivar == some_value`
  - `catch <ErrorName>` which can be used to break when some error is raised.
  - Ways to quickly examine local context (`info` command, using the vscode UI's side panels, etc.)
  
  Of course this stuff can get super deep, and I'd rather not overwhelm people. It's probably important to keep in mind that this lesson is happening before learners have even done Tic-Tac-Toe. But I do find myself using the above things pretty often.
  
- I agree that `binding.irb` is probably a good way to ease people in to interactive debugging.

- Also agree that there's room for two lessons here. There's just a lot going on between introducing the callstack, `p/puts` debugging, interactive debugging, setting up vscode debugger etc. Setting up vscode debugging definitely deserves more real estate than it currently has. It's a very powerful tool, and learners are still at the stage of being a bit intimidated by things like the whole `launch.json` setup. I get the feeling that some people never get it working and just stick with `pry`, which is a bit of a shame I think.

**Comment by KevinMulhern** [comment]
Hey @ChargrilledChook, @JoshDevHub, can we close [this related issue](https://github.com/TheOdinProject/curriculum/issues/25016) now that we're moving ahead with this different approach?

**Comment by ChargrilledChook** [comment]
@KevinMulhern Good call mate, sorted :+1: 

@JoshDevHub great write up, I think we're very much aligned here.

As a way of breaking this up a bit (and letting me find those drafts :sweat_smile: ) would you able to take on part one of the suggested approach and streamline the first lesson? 

If you want to do a straight one to one swap (removing references to pry and the VSCode debugger) that would be ok, but if you wanted to expand on some of the more conceptual points that could be useful too.  

This could be a section that would really benefit from having its own RSpec style playground repo, although that might be blowing the scope out of this particular issue a little - but happy to discuss how that might look.

**Comment by crespire** [comment]
I would be happy to help out on the increased scope playground idea once it's ready for work.

**Comment by ChargrilledChook** [comment]
@JoshDevHub Yep, perfect :ok_hand:

**Comment by radanskoric** [comment]
Hi, I was wondering what is the state of this ticket as I would like to contribute.

I was browsing Ruby content and noticed on the debugging lesson that it still has just byebug content and doesn't mention the new debug gem. Since I have experience with both, I thought I would contribute to the article, update it to be about the debug gem instead of pry-byebug. 

Then I checked PRs and found https://github.com/TheOdinProject/curriculum/pull/25276 which lead me to this issue. 

I was thinking of writing a section about the debug gem that would on top of the changes from the linked PR but I'm not entirely sure if tha fits within the overarching plan of this ticket?

What would be the best way forward for me to contribute to the debugging lesson?
